// @securityscorecard/cli/build/commands/simulate.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var ora_1 = __importDefault(require("ora"));
var apps_1 = __importDefault(require("../services/apps"));
var inquires = __importStar(require("../utils/inquires"));
var rc = __importStar(require("../utils/rc"));
var helpers_1 = require("../utils/helpers");
var logger_1 = require("../utils/logger");
var inquires_1 = require("../utils/inquires");
var json_1 = require("../utils/json");
var operationGetSubscriptions = 'Getting Rules';
var operationSimulateRule = 'Simulating Rule';
var operationLoadingFile = 'Loading event from file';
var message = 'In which environment do you want to trigger the event?';
var simulate = function (args) { return __awaiter(void 0, void 0, void 0, function () {
    var env, _a, token, filePath, fakeEvent, loadingFile, gettingSubscriptions, subscriptions, err_1, availableSubscriptions, availableRules, selection, rule, rawEvent, err_2, simulating, event_1, err_3;
    var _b, _c;
    return __generator(this, function (_d) {
        switch (_d.label) {
            case 0:
                if (!args.environment) return [3 /*break*/, 1];
                _a = 'production';
                return [3 /*break*/, 3];
            case 1: return [4 /*yield*/, inquires.askEnvironment(message)];
            case 2:
                _a = (_d.sent()).environment;
                _d.label = 3;
            case 3:
                env = _a;
                token = (_b = rc.readRc(helpers_1.SSC_RC_PATH)[env]) === null || _b === void 0 ? void 0 : _b.token;
                filePath = args.file;
                if (filePath) {
                    loadingFile = (0, ora_1["default"])(operationLoadingFile).start();
                    try {
                        fakeEvent = (0, json_1.load)(filePath);
                        loadingFile.succeed();
                    }
                    catch (err) {
                        loadingFile.fail();
                        (0, logger_1.log)((0, logger_1.error)('Invalid file arg'));
                        (0, logger_1.log)((0, logger_1.error)(err.message));
                        return [2 /*return*/];
                    }
                }
                gettingSubscriptions = (0, ora_1["default"])(operationGetSubscriptions).start();
                if (!token) {
                    gettingSubscriptions.fail();
                    (0, logger_1.log)((0, logger_1.error)("No token found for " + env + ", you must set your config first"));
                    (0, logger_1.log)((0, logger_1.error)('ie.: ssc config'));
                    return [2 /*return*/];
                }
                subscriptions = {};
                _d.label = 4;
            case 4:
                _d.trys.push([4, 6, , 7]);
                return [4 /*yield*/, apps_1["default"].subscriptions({ token: token, env: env })];
            case 5:
                subscriptions = (_d.sent());
                return [3 /*break*/, 7];
            case 6:
                err_1 = _d.sent();
                gettingSubscriptions.fail();
                (0, logger_1.log)((0, logger_1.error)('API error'));
                (0, logger_1.log)((0, logger_1.error)("Error: " + err_1.message));
                return [2 /*return*/];
            case 7:
                availableSubscriptions = ((_c = subscriptions === null || subscriptions === void 0 ? void 0 : subscriptions.entries) === null || _c === void 0 ? void 0 : _c.filter(function (s) { return !s.delivery.workflow.hidden; })) || [];
                if (!availableSubscriptions.length) {
                    gettingSubscriptions.fail();
                    (0, logger_1.log)((0, logger_1.error)('\nNo rules found!'));
                    (0, logger_1.log)((0, logger_1.info)('Creation of new rules could be done on https://platform.securityscorecard.io/#/my-settings/rules'));
                    return [2 /*return*/];
                }
                gettingSubscriptions.succeed();
                availableRules = availableSubscriptions.map(function (s) { return ({
                    id: s.id,
                    // @ts-ignore
                    type: s.event_type,
                    name: s.delivery.workflow.name
                }); });
                return [4 /*yield*/, (0, inquires_1.askEventToSimulate)(availableRules.map(function (e) { return ({
                        name: e.name
                    }); }))];
            case 8:
                selection = _d.sent();
                rule = availableRules.find(function (r) { return r.name === selection.rule; });
                if (!!fakeEvent) return [3 /*break*/, 12];
                _d.label = 9;
            case 9:
                _d.trys.push([9, 11, , 12]);
                return [4 /*yield*/, (0, inquires_1.askForFakeEvent)()];
            case 10:
                rawEvent = _d.sent();
                fakeEvent = JSON.parse(rawEvent);
                return [3 /*break*/, 12];
            case 11:
                err_2 = _d.sent();
                (0, logger_1.log)((0, logger_1.error)('Invalid fake event'));
                return [2 /*return*/];
            case 12:
                simulating = (0, ora_1["default"])(operationSimulateRule).start();
                (0, logger_1.log)((0, logger_1.info)('\nSimulating:\n', JSON.stringify(fakeEvent)));
                _d.label = 13;
            case 13:
                _d.trys.push([13, 15, , 16]);
                return [4 /*yield*/, apps_1["default"].events({
                        token: token,
                        env: env,
                        ruleId: rule.id,
                        type: rule.type,
                        event: fakeEvent
                    })];
            case 14:
                event_1 = _d.sent();
                simulating.succeed();
                if (event_1.received)
                    (0, logger_1.log)((0, logger_1.info)('🦾 Rule event triggered, check your app logs!'));
                return [3 /*break*/, 16];
            case 15:
                err_3 = _d.sent();
                simulating.fail();
                (0, logger_1.log)((0, logger_1.error)('Something went wrong.', err_3));
                return [3 /*break*/, 16];
            case 16: return [2 /*return*/];
        }
    });
}); };
exports["default"] = simulate;
